kubectl prunev2: Refactor the applyset to be more reusable

justinsb opened the pull request:
This enables sharing with diff.

/kind cleanup

```release-note
NONE
```

justinsb committed:
kubectl prunev2: Refactor the applyset to be more reusable

This enables sharing with diff.

KnVerey commented on staging/src/k8s.io/kubectl/pkg/cmd/apply/applyset_pruner.go:
```diff
@@ -31,101 +32,119 @@ import (
 
-func (p *applySetPruner) delete(ctx context.Context, namespace string, name string, mapping *meta.RESTMapping) error {
-	return runDelete(ctx, namespace, name, mapping, p.dynamicClient, p.cascadingStrategy, p.gracePeriod, p.dryRunStrategy == cmdutil.DryRunServer)
+func (a *ApplySet) delete(ctx context.Context, dynamicClient dynamic.Interface, namespace string, name string, mapping *meta.RESTMapping, o *ApplySetDeleteOptions) error {
```
nit: would this make more sense as a method on `ApplySetDeleteOptions`? It uses three of its values, and doesn't interact with `ApplySet` at all. Or maybe just call `runDelete` directly from `deleteObjects` unless I'm missing some value in the indirection here.

KnVerey commented on staging/src/k8s.io/kubectl/pkg/cmd/apply/applyset_pruner.go:
```diff
@@ -31,101 +32,119 @@ import (
+
+		if opt.DryRunStrategy != cmdutil.DryRunClient {
+			if err := a.delete(ctx, dynamicClient, namespace, name, mapping, opt); err != nil {
+				return fmt.Errorf("pruning %v %s/%s: %w", pruneObject.Mapping.GroupVersionKind.String(), namespace, name, err)
```
nit: maybe make a `pruneObject.String()` than could prevent the slash for non-namespaced objects? (e.g. `Kind=Namespace /bar` in the test)

KnVerey commented on staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go:
```diff
@@ -1027,11 +1016,7 @@ func (o *ApplyOptions) PrintAndPrunePostProcessor() func() error {
-					return err
-				}
-				if err := pruner.pruneAll(ctx, o.ApplySet); err != nil {
+				if err := o.ApplySet.Prune(ctx, o); err != nil {
```
Should `o.ApplySet.UpdateParent` be merged into `o.ApplySet.Prune` as well (and then UpdateParent made private)? IIRC I originally had it that way but moved it out because the parent update went in before pruning was implemented at all. I don't think there's a conceivable case where pruning was completed but parent update should not be attempted?

KnVerey commented on staging/src/k8s.io/kubectl/pkg/cmd/apply/applyset.go:
```diff
@@ -457,3 +459,58 @@ func ParseApplySetParentRef(parentRefStr string, mapper meta.RESTMapper) (*Apply
+
+// FindObjectsToPrune returns the list of objects that will be pruned.
+// Calling this instead of Prune can be useful for dry-run / diff behaviour.
+func (a *ApplySet) FindObjectsToPrune(ctx context.Context, dynamicClient dynamic.Interface, visitedUids sets.Set[types.UID]) ([]PruneObject, error) {
```
Why have `FindObjectsToPrune` that is public wrapping the very similarly named `findAllObjectsToPrune` that is private?

justinsb replied:
Good point, removed the extra layer.

KnVerey commented on staging/src/k8s.io/kubectl/pkg/cmd/apply/applyset.go:
```diff
@@ -457,3 +459,58 @@ func ParseApplySetParentRef(parentRefStr string, mapper meta.RESTMapper) (*Apply
 }
+
+// Prune deletes any objects from the apiserver that are no longer in the applyset.
+func (a *ApplySet) Prune(ctx context.Context, o *ApplyOptions) error {
```
Same question as below about having a `Prune` and a `pruneAll` on the same object (though this one does more than just wrap). Also, now that applySetPruner isn't a separate object, should that file's content move here? Or if we're collecting the pruning related functionality in that file, shouldn't this and the method below be there?

KnVerey replied:
Separate package in a later refactor SGTM!

KnVerey commented:
/test pull-kubernetes-unit
/kind cleanup
/approve
/triage accepted
/priority important-longterm

justinsb commented:
Thanks for review - addressed the nits (which I agree with, though I think we should do the refactor in a separate PR once we have this code shared with diff)

KnVerey commented:
/lgtm
/unhold

KnVerey commented:
/lgtm

KnVerey commented:
/lgtm
